### PR TITLE
Check edge cases for automounted configmaps/secrets

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Checks some edge cases in automounted secrets and volumes that can cause workspace startup to hang
1. Fail startup if the deployment we're creating is invalid
2. Check for collision in automounted volumes:
    * automounted volume collides with DevWorkspace volume on name
    * automounted volume collides with DevWorkspace volume on mountPath
    * automounted volumes collide on mountPath

Also prefixes automounted volumes with `automount-[secret|configmap]` to avoid collisions with each other and with DevWorkspace volumes (e.g. you can't automount a configmap named "project" currently).

**Note**: prefixing volume names as above can lead to issues, so it may be worth it to drop the last commit in this PR. If a secret/configmap's name is very long and the prefix pushes it over 63 chars, workspace start will fail.

### What issues does this PR fix or reference?
Minor issue I happened to run into while checking https://github.com/devfile/devworkspace-operator/issues/455 where my secrets shared a mountPath and no error was shown (except in logs)

### Is it tested? How?
Here comes a long testing doc:
1. Store the DevWorkspace we'll use for testing in /tmp/
    ```yaml
    cat > /tmp/test-dw.yaml <<EOF
    kind: DevWorkspace
    apiVersion: workspace.devfile.io/v1alpha2
    metadata:
      name: theia-next
    spec:
      started: true
      template:
        components:
          - name: test-container
            container:
              image: quay.io/libpod/busybox:latest
              volumeMounts:
                - name: testing-volume
                  path: /test-path
          - name: theia
            plugin:
              uri: https://che-plugin-registry-main.surge.sh/v3/plugins/eclipse/che-theia/next/devfile.yaml
          - name: testing-volume
            volume: {}
    EOF
    ```
2. Create secrets and configmaps for testing
    ```bash
    kubectl create secret generic test
    kubectl create configmap test
    kubectl patch secret test --type merge -p \
      '{
        "metadata": {
          "labels": {
            "controller.devfile.io/mount-to-devworkspace": "true"
          }
        }
      }'
    kubectl patch cm test --type merge -p \
      '{
        "metadata": {
          "labels": {
            "controller.devfile.io/mount-to-devworkspace": "true"
          }
        }
      }'
    ```
3. Case 1: automounted volumes collide on mountPath
    ```bash
    kubectl patch secret test --type merge -p \
      '{
        "metadata": {
          "annotations": {
            "controller.devfile.io/mount-path": "/collision"
          }
        }
      }'
    kubectl patch cm test --type merge -p \
      '{
        "metadata": {
          "annotations": {
            "controller.devfile.io/mount-path": "/collision"
          }
        }
      }'
    kubectl apply -f /tmp/test-dw.yaml
    kubectl get dw -w
    ```
    (should fail with meaningful error message)
4. Case 2: automounted volume and DevWorkspace volume collide on mountPath
    ```bash
    kubectl delete dw theia-next
    kubectl patch cm test --type merge -p \
      '{
        "metadata": {
          "annotations": {
            "controller.devfile.io/mount-path": "/test-path"
          }
        }
      }'
    kubectl apply -f /tmp/test-dw.yaml
    kubectl get dw -w
    ```
    (should fail with meaningful error message)

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v7-devworkspaces-operator-e2e, v7-devworkspace-happy-path` to trigger)
    - [ ] `v7-devworkspaces-operator-e2e`: DevWorkspace e2e test
    - [ ] `v7-devworkspace-happy-path`: DevWorkspace e2e test

### CODE REVIEW

File: controllers/workspace/provision/automount.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+		}
+	}
+
+	// Check that automounted volume mountPaths do not conflict with existing mountPaths in any container
+	for _, podAddition := range base {
+		for _, container := range podAddition.Containers {
+			for _, vm := range container.VolumeMounts {
+				if conflict, exists := automountVolumeMountsByMountPath[vm.MountPath]; exists {
+					return fmt.Errorf("DevWorkspace volume %s in container %s has same mountpath as auto-mounted volume from %s",
+						getVolumeDescriptionFromVolumeMount(vm, base), container.Name, getVolumeDescriptionFromVolumeMount(conflict, automount))
```
Comment (sleshchenko) [inline controllers/workspace/provision/automount.go:205]: Do you have a check on the previous stage that different volumes are not mounted on the same path? Probably not or at least I haven't found it easily.
If we don't - please create an issue to validate flattened devfile on any kind of conflicts

**Review by sleshchenko — APPROVED** [review]
I haven't tested but changes look meaningful.

Ideally, we should have unit tests for covered use-cases, but it's not mandatory.

File: controllers/workspace/provision/automount.go (hunk 1 above)
Comment (amisevsk) [reply controllers/workspace/provision/automount.go:205]: Good point -- we do some checking but it's not consistent. Created https://github.com/devfile/devworkspace-operator/issues/497

**Comment by amisevsk** [comment]
> Ideally, we should have unit tests for covered use-cases, but it's not mandatory.

I'll look into this -- I've generally been avoidant of adding tests for things that require the k8s API, so maybe it's time to dive into fakeClient :)

**Review by JPinkney — APPROVED** [review]
Changes make sense to me :+1:

File: controllers/workspace/provision/automount_test.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+					mountPath:  "/test/mount",
+					volumeType: secretVolumeType,
+				},
+				{
+					name:       "testVolume2",
+					mountPath:  "/test/mount",
+					volumeType: configMapVolumeType,
+				},
+			},
+			errRegexp: "auto-mounted volumes from configmap testVolume2 and secret testVolume1 have the same mount path",
```
Comment (sleshchenko) [inline controllers/workspace/provision/automount_test.go:119]: Maybe it makes sense to take in quotes objects names?

Comment (amisevsk) [reply controllers/workspace/provision/automount_test.go]: Done :)

File: controllers/workspace/provision/automount_test.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+				},
+			},
+			automountPodAdditions: []volumeDesc{
+				{
+					name:       "testVolume",
+					mountPath:  "/collision/path",
+					volumeType: secretVolumeType,
+				},
+			},
+			errRegexp: fmt.Sprintf("DevWorkspace volume baseVolume in container %s has same mountpath as auto-mounted volume from secret 'testVolume'", testContainerName),
```
Comment (sleshchenko) [inline controllers/workspace/provision/automount_test.go:103]:
```suggestion
			errRegexp: fmt.Sprintf("DevWorkspace volume 'baseVolume' in container %s has same mountpath as auto-mounted volume from secret 'testVolume'", testContainerName),
```

File: controllers/workspace/provision/automount_test.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+				},
+			},
+			automountPodAdditions: []volumeDesc{
+				{
+					name:       "baseVolume",
+					mountPath:  "/configmap/mount",
+					volumeType: configMapVolumeType,
+				},
+			},
+			errRegexp: "DevWorkspace volume baseVolume conflicts with automounted volume from configmap 'baseVolume'",
```
Comment (sleshchenko) [inline controllers/workspace/provision/automount_test.go:85]:
```suggestion
			errRegexp: "DevWorkspace volume 'baseVolume' conflicts with automounted volume from configmap 'baseVolume'",
```

**Review by sleshchenko — APPROVED** [review]

**Pull request merged**